### PR TITLE
Add normalized power metrics

### PR DESCRIPTION
**Pull request by tjablin**

### CODE REVIEW

**Comment by DilipSequeira** [comment]
@tjablin for clarification - this is a number that appears only in the power tables, and is calculated as measured performance / measured energy?

**Comment by rnaidu02** [comment]
Power WG (Sachin) would like to discuss this item with Tom Jablin.

**Review by DilipSequeira — APPROVED** [review]
lgtm

**Comment by s-idgunji** [comment]
Power WG will propose the appropriate text for the table and the normalized metric added  (if this normalized metric is agreed as an addition by Inference WG)

**Comment by psyhtest** [comment]
I believe that the [Results Guidelines](https://github.com/mlcommons/policies/blob/master/RESULTS_GUIDELINES.md#mlperf-results-must-include-a-detailed-footnote) are here to help. In particular, quoting the MLPerf result ID automatically identifies the system under test, the software, etc. Including the normalized Pe1/Po1 metric into the result's row will prevent its misuse with another Pe2 result from another row (with another result ID).

The only problem I see is that someone can take the normalized metric for one workload and use it with another workload **in the same row**, but given that samples/Joule may be very different for different workloads (e.g. 200 for ResNet, 10 for BERT) it's not a big concern.

**Comment by psyhtest** [comment]
The below figure has appeared on [NVIDIA's blog](https://developer.nvidia.com/blog/getting-the-best-performance-on-mlperf-inference-2-0/) with the following caption:

> MLPerf v2.0 Inference Edge Closed and Edge Closed Power; Performance/Watt from MLPerf results for respective submissions for Data Center and Edge, Offline Throughput, and Power. NVIDIA Xavier AGX Xavier: 1.1-110 and 1.1-111 | Jetson AGX Orin: 2.0-140 and 2.0-141.

edge-performance

For ResNet50, the claimed Performance improvement of ~3x could have been taken either from (6,138.84 / 2,039.11 = 3.01) or  (4,750.26 / 1,506.53 = 3.15).

System | Submission | Performance | Power | Performance / Power
-|-|-|-|-
Xavier | 1.1-110 |  2,039.11 | N/A | N/A
Xavier (MaxQ) | 1.1-111 |  1,506.53 | 25.24 | 59.67
Orin | 2.0-140 | 6,138.84 | N/A | N/A
Orin (MaxQ) | 2.0-141 | 4,750.26 | 42.42 | 111.98

**Comment by psyhtest** [comment]
Throwing another system into the mix ((Foxconn Gloria](http://www.ingrasys.com/ea5165n) powered by [Qualcomm AI100 DM.2e](https://www.qualcomm.com/products/technology/processors/cloud-artificial-intelligence/cloud-ai-100)) on ResNet50:

System | Submission | Performance | Power | Performance / Power
-|-|-|-|-
Xavier | 1.1-110 |  2,039.11 | N/A | N/A
Xavier (MaxQ) | 1.1-111 |  1,506.53 | 25.24 | 59.67
Orin | 2.0-140 | 6,138.84 | N/A | N/A
Orin (MaxQ) | 2.0-141 | 4,750.26 | 42.42 | 111.98
Gloria | 2.0-129 | 7,309.24 | 24.34 | 300.30

which of the comparisons are valid:
- Gloria is 2.7x (= 300 / 112) more energy efficient than Orin (MaxQ) and 5x (= 300 / 60) more energy efficient than Xavier (MaxQ).
- Gloria is 1.5x (= 7,309 / 4,750) faster than Orin (MaxQ) and 4.9x (= 7,309 / 1,506) faster than Xavier (MaxQ).
- Gloria is 1.2x (= 7,309 / 6,139) faster than Orin (non-MaxQ) and 3.6x (= 7,309 / 2,039) faster than Xavier (non-MaxQ).

**Comment by s-idgunji** [comment]
> The below figure has appeared on [NVIDIA's blog](https://developer.nvidia.com/blog/getting-the-best-performance-on-mlperf-inference-2-0/) with the following caption:
> 
> > MLPerf v2.0 Inference Edge Closed and Edge Closed Power; Performance/Watt from MLPerf results for respective submissions for Data Center and Edge, Offline Throughput, and Power. NVIDIA Xavier AGX Xavier: 1.1-110 and 1.1-111 | Jetson AGX Orin: 2.0-140 and 2.0-141.
> 
> edge-performance
> 
> For ResNet50, the claimed Performance improvement of ~3x could have been assumed to be taken from either non-MaxQ (6,138.84 / 2,039.11 = 3.01) or MaxQ (4,750.26 / 1,506.53 = 3.15).
> 
> System	Submission	Performance	Power	Performance / Power
> Xavier	1.1-110	2,039.11	N/A	N/A
> Xavier (MaxQ)	1.1-111	1,506.53	25.24	59.67
> Orin	2.0-140	6,138.84	N/A	N/A
> Orin (MaxQ)	2.0-141	4,750.26	42.42	111.98
> But for BERT-99, the claimed Performance improvement of ~5x could have only come from non-MaxQ (476.34 / 96.73 = 4.92), not from MaxQ (394.33 / 61.35 = 6.43).
> 
> System	Submission	Performance	Power	Performance / Power
> Xavier	1.1-110	96.73	N/A	N/A
> Xavier (MaxQ)	1.1-111	61.35	18.89	3.25
> Orin	2.0-140	476.34	N/A	N/A
> Orin (MaxQ)	2.0-141	394.33	53.59	7.36
> Therefore, it appears that the normalized metric is used out of context here, contrary to [the last comment](https://github.com/mlcommons/inference_policies/pull/227#issuecomment-949942131). I think this would be cleaner to have two separate figures:
> 
> 1. Showing the improvement of Performance from Xavier (non-MaxQ) to Orin (non-MaxQ) with the caption:
> 
> > NVIDIA Xavier AGX Xavier: 1.1-110 | Jetson AGX Orin: 2.0-140.
> 
> 2. Showing the improvement of Performance / Power (and possibly Performance as well) from Xavier (MaxQ) to Orin (MaxQ) with the caption:
> 
> > NVIDIA Xavier AGX Xavier: 1.1-111 | Jetson AGX Orin: 2.0-141.
> 
> I'm not saying the NVIDIA figure violates any rule today, but we should be more explicit about how to make such comparisons in the future.

Hi @psyhtest  - We discussed this in the Power WG and from the discussions with @anirban-ghosh  in the meeting,  that the table is fine.   On discussing with @DilipSequeira  and/or @georgelyuan, if there are any questions please bring up in the Power WG next week.  Thanks.

**Comment by psyhtest** [comment]
>  We discussed this in the Power WG and from the discussions with @anirban-ghosh in the meeting, that the table is fine.

I believe Tejus had the same reservation as me. George agrees that this could have been more explicit. Let's revisit in the next meeting.

**Comment by psyhtest** [comment]
Let P<sub>_id_</sub> be the Performance (samples per second) and W<sub>_id_</sub> be the Power of submission number _id_, and R<sub>_id_</sub> to be their ratio, i.e. the normalized metric. For  example, for ResNet50:

System | Submission | Performance | Power | Performance / Power
-|-|-|-|-
Xavier | 1.1-110 |  P<sub>_1.1-110_</sub>=2,039.11 | W<sub>_1.1-110_</sub>=N/A | R<sub>_1.1-110_</sub>=N/A
Xavier (MaxQ) | 1.1-111 |  P<sub>_1.1-111_</sub>=1,506.53 | W<sub>_1.1-111_</sub>=25.24 | R<sub>_1.1-111_</sub>=59.67
Orin | 2.0-140 | P<sub>_2.0-140_</sub>=6,138.84 | W<sub>_2.0-140_</sub>=N/A | R<sub>_2.0-140_</sub>=N/A
Orin (MaxQ) | 2.0-141 | P<sub>_2.0-141_</sub>=4,750.26 | W<sub>_2.0-141_</sub>=42.42 | R<sub>_2.0-141_</sub>=111.98

The figure effectively shows (P<sub>_2.0-140_</sub> / P<sub>_1.1-110_</sub> = 3.01) as "Perf" bars, and (R<sub>_2.0-141_</sub> / R<sub>_1.1-111_</sub> = 1.88) as "Perf/Watt" bars next to each other.

But the latter is actually (P<sub>_2.0-141_</sub> / W<sub>_2.0-141_</sub>) / (P<sub>_1.1-111_</sub> / W<sub>_1.1-111_</sub>) = (P<sub>_2.0-141_</sub> / P<sub>_1.1-111_</sub>) * (W<sub>_1.1-111_</sub> / W<sub>_2.0-141_</sub>) = 3.15 * 0.59. 

In summary, the adjacent "Perf" and "Perf/Watt" bars are based on the ratio of different Performance values.

**Comment by psyhtest** [comment]
We know the ratio of (W<sub>_1.1-111_</sub> / W<sub>_2.0-141_</sub> = 0.59) because W<sub>_1.1-111_</sub> and W<sub>_2.0-141_</sub> were measured and reported by NVIDIA. 

@s-idgunji As you said many times yourself, the Perf/Watt ratio cannot be assumed constant across different modes such as MaxN and MaxQ.

Unfortunately, we do not know the MaxN ratio (W<sub>_1.1-110_</sub> / W<sub>_2.0-140_</sub>). But, hypothetically, if (W<sub>_1.1-110_</sub>= 50) and (W<sub>_2.0-140_</sub> = 70), this ratio is 0.71. Conversely, if (W<sub>_1.1-110_</sub>= 50) and the ratio is assumed the same as the MaxQ ratio (0.59), (W<sub>_2.0-140_</sub> = 70) = 85. Without taking any Orin measurements myself, I find the former hard to believe.

**Comment by s-idgunji** [comment]
=============================================================================
>> @s-idgunji As you said many times yourself, the Perf/Watt ratio cannot be assumed constant across different modes such as MaxN and MaxQ.

And nothing in the table is breaking this ask.  As far as I can tell,  there is no Watt with the  Perf only submission and neither is the Max Perf being used that is the case in Perf/W comparison.  The rule is that any Perf/W can use  Perf with associated Watt and no other. And one can only refer to the measured Perf/W in the comparison.

If normalized data in the NVIDIA table needs to be presented separately , footnoted clearly  to avoid confusion, and you have suggestions,  please reach out to NVIDIA directly.  If on the other hand you think that the ratios used are incorrect for Perf/W , i.e  incorrect Perf Value with incorrect Power Value , then please point out in our next WG. 

And do you want to propose guidelines on how Perf/W be normalized and results be presented by submitters? That can be discussed separately as well.

But, we want to be careful that eventually Perf/W comparisons can also lead to Perf comparisons.  For e.g.  how can submitters compare  "Perf only"  Perf data vs    "Perf, Power"  Perf data .  Is that a valid comparison?  Some submitter could do that.  So you can see that this can become increasingly complicated as submissions can cover energy efficient points to  max performance points and so on.  

I also want to point out that this PR is specifically about adding a foot note in MLC externally visible tables on our MLCommons website that should state that any normalized Perf/W data is only valid for that particular entry and no other.

**Comment by DilipSequeira** [comment]
[Just catching up on this thread.] 

Suppose that MLPerf had additional metrics (say, DRAM usage, and system noise in dB) with entries optimized for each using different SW configuration options. Would it be legitimate to present a comparison of two systems across the scenarios capturing all of these things in a single chart? I think it's useful to be able to do so, and we should think about what criteria we want to impose in such cases.

I would say, at minimum:

* each measurement point the graph is apples-to-apples (so in the case of dB, it should compare the noise-optimized configurations, DRAM should compare the DRAM-optimized configurations, same for unconstrained perf and power efficiency.)
* the measurement axis can be sensibly numbered (which in practice I think requires normalizing to one system so that the values are dimensionless.)
* footnotes should capture which comparison points are being taken from which lines in the results tables.

AFAICT our blog post conforms to the first two of these, and is consistent with the spirit of the rules as well as the letter. It doesn't conform to the third, and hence is somewhat imprecise about what's being presented. I think it would be reasonable to refine the requirements around footnoting so that it's clearer how such tables should be annotated in future rounds to prevent any possible confusion.

**Comment by arjunsuresh** [comment]
I think this PR is useful to be merged.  Joules/stream can be added as an additional column in the results table as it makes life so easier for any user seeing the table. The following data is taken from the 3.0 results and is for resnet50 model. The first 4 columns are from the results table and I'm not sure a normal user can do a simple comparison here based on the values shown. But the last column - Joules per sample gives a straight away comparison and gives a useful metric to the user and this also gives a power efficiency comparison across the 4 different scenarios. 

<google-sheets-html-origin><style type="text/css"></style>
Perf | result log | Testing Watts | Power/Energy reported | Joules/stream
-- | -- | -- | -- | --
9801.8 | closed/Qualcomm/results/gloria_highend-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 31.22771 | 31.22771 | 313.8814854

6810.02 | closed/Qualcomm/results/heimdall-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 23.77151 | 23.77151 | 286.4782254
7025.39 | closed/Qualcomm/results/gloria-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 24.758 | 24.758 | 283.7624202
6792.57 | closed/Krai/results/eb6-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 24.1111 | 24.1111 | 281.7196229
7041.3 | closed/Krai/results/rb6-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 25.28519 | 25.28519 | 278.4752656
169564 | closed/Qualcomm/results/r282_z93_q8e-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 701.30912 | 701.30912 | 241.7821117
145881 | closed/Qualcomm/results/r282_z93_q8e-qaic-v1.8.3.7-aic100/resnet50/server/performance | 641.53017 | 641.53017 | 227.395385
333078 | closed/Qualcomm/results/g292_z43_q16e-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 1565.21104 | 1565.21104 | 212.8006968
106199 | closed/Qualcomm/results/r282_z93_q5e-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 516.55931 | 516.55931 | 205.5891704
2.203008 | closed/Qualcomm/results/gloria_highend-qaic-v1.8.3.7-aic100/resnet50/multistream/performance | 18.02134 | 4.962644524 | 201.5054665
290034 | closed/Qualcomm/results/g292_z43_q16e-qaic-v1.8.3.7-aic100/resnet50/server/performance | 1467.62883 | 1467.62883 | 197.6208113
361537 | closed/Qualcomm/results/g292_z43_q18e-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 1829.44855 | 1829.44855 | 197.620753
345043 | closed/Qualcomm/results/g292_z43_q18e-qaic-v1.8.3.7-aic100/resnet50/server/performance | 1751.46667 | 1751.46667 | 197.0023215
79193.9 | closed/HPE/results/e920d_q4_std-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 426.60129 | 426.60129 | 185.639148
94615.2 | closed/Dell/results/r7515_q4_pro-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 511.90636 | 511.90636 | 184.829116
2.05857 | closed/Qualcomm/results/heimdall-qaic-v1.8.3.7-aic100/resnet50/multistream/performance | 21.16229 | 5.445506916 | 183.6376329
93750.3 | closed/Dell/results/r7515_q4_pro-qaic-v1.8.3.7-aic100/resnet50/server/performance | 516.45558 | 516.45558 | 181.526357
2.128724 | closed/Krai/results/rb6-qaic-v1.8.3.7-aic100/resnet50/multistream/performance | 20.96114 | 5.577560223 | 179.2898615
2.161895 | closed/Qualcomm/results/gloria-qaic-v1.8.3.7-aic100/resnet50/multistream/performance | 21.06942 | 5.693734219 | 175.6316613
2.158166 | closed/Krai/results/eb6-qaic-v1.8.3.7-aic100/resnet50/multistream/performance | 22.20422 | 5.990049083 | 166.9435402
353626 | closed/NVIDIA/results/H100-PCIe-80GBx8_TRT_MaxQ/resnet50/offline/performance | 2219.57741 | 2219.57741 | 159.3213187
3450.66 | closed/NVIDIA/results/Orin_TRT_MaxQ/resnet50/offline/performance | 22.65936 | 22.65936 | 152.2840892
24768 | closed/Lenovo/results/se350_q1_pro-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 168.67725 | 168.67725 | 146.8366362
2255.01 | closed/SiMa/results/davinci_dualm2/resnet50/offline/performance | 16.1066 | 16.1066 | 140.0053394
46417.2 | closed/Lenovo/results/se450_q4_lite-qaic-v1.8.3.7-aic100/resnet50/offline/performance | 345.90388 | 345.90388 | 134.191036
3.708546 | closed/SiMa/results/davinci_dualm2/resnet50/multistream/performance | 16.24655 | 7.531384752 | 132.7777073
0.514955 | closed/Qualcomm/results/heimdall-qaic-v1.8.3.7-aic100/resnet50/singlestream/performance | 15.35772 | 7.908534703 | 126.4456739

**Comment by arjunsuresh** [comment]
Adding the link to the full results with unified power metric (inference per Joules) [here](https://docs.google.com/spreadsheets/u/3/d/e/2PACX-1vRGM7md3feKd9L9rjBcQL0xETrtIXOnbntlBVkbsxEGyoBXhmc2W5IBuNkdBVvp0f64C5GnaBm7-4HI/pubhtml)

**Comment by mrmhodak** [comment]
Needs more discussion, need opinion from Power

**Comment by s-idgunji** [comment]
> Needs more discussion, need opinion from Power

We can discuss this.  We had already agreed to adding energy metrics.  What is this new update ?

**Comment by s-idgunji** [comment]
> > Needs more discussion, need opinion from Power
> 
> We can discuss this. We had already agreed to adding energy metrics. What is this new update ?

@mrmhodak  - We had a discussion in the Power WG meeting.  

1)  it's unclear how the samples/J is calculated (it is derive with unambiguously relating it to Performance , because performance has no metric associated  , nor from Power )

2) Media, Reviewers viewing the table are already using Perf (measured) and Power (measured) data and then deriving the metric, so adding a 3rd column is redundant (no new information , other than ease of reading)

3) Krai claims that there have been issues with having the measured metrics to derive and energy efficiency . There are some links/websites mentioned.  Krai will provide the data/evidence for such issues

4) Measured metrics must be always in results table. 

5) The metrics to be reported as captured in the Inference Power methodology document in MLPerf Power area clearly states that for reporting   :  if the the measurement is a rate  ( Queries/Sec or Samples/Sec) , the associated electrical metric is  Power  (Joules/Sec)  , i.e a rate  .   If the measurement is time or latency , the associated electrical metric is Energy (Joules or mJ)

6) It's even more complicated/ambiguous for single stream and multi-stream given current reporting :  the performance metric is  latency (no reference to throughput and what specifically).  The electrical metric is  Energy (which is 1:1 corresponding metric).  The suggestion of samples/J has no way of deriving how it came about from the current present metrics { Latency , Energy}  .  Needs further discussion.

The next steps are to follow on in the next instance  (had to cover another 3.1 submission topic) 

Sachin

**Comment by mrmhodak** [comment]
Closing, no longer relevant with new spreadsheet coming soon.
